File: application/app/domain/model/filters/WhitelistHosts.scala
package domain.model.filters

import common.validation.constraints.TraversableSize
import domain.model.{Filter, SetValue}
import javax.validation.constraints.NotNull

import scala.annotation.meta.field

case class WhitelistHosts(
                           @(NotNull @field) @(TraversableSize @field)(max = 255) whitelist: Set[String]
                         ) extends Filter {
  override val name: String = Filter.WHITELIST_HOSTS
  override val value: SetValue = SetValue(whitelist)
  override val filter: Map[String, String] => Boolean = headers => headers.contains("Host") && whitelist.contains(headers("Host"))
}

File: ws-gateway-domain/src/main/java/com/cosmin/wsgateway/domain/filters/WhitelistHosts.java
package com.cosmin.wsgateway.domain.filters;

import com.cosmin.wsgateway.domain.Filter;
import java.util.Map;
import java.util.Set;
import lombok.Value;

@Value
public class WhitelistHosts implements Filter<Set<String>> {
    private final Set<String> hosts;

    @Override
    public String name() {
        return "whitelistHosts";
    }

    @Override
    public Set<String> value() {
        return hosts;
    }

    @Override
    public boolean isAllowed(Map<String, String> requestHeaders) {
        if (hosts.isEmpty()) {
            return true;
        }
        return requestHeaders.containsKey("Host") && hosts.contains(requestHeaders.get("Host"));
    }
}
